backstory/src/demo/java/com/google/sps/servlets/PerspectiveServlet.java

```diff
@@ -39,11 +45,94 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     response.setContentType("application/json;");
     Gson gson = new Gson();
 
-    String[] cannedOutput = new String[0];
+    // will hold text to analyze
+    String text;
 
-    // write canned output to servlet
-    String json = gson.toJson(cannedOutput);
+    String json = request.getReader().readLine();
 
-    response.getWriter().println(json);
+    // get the text from the JSON & handle error if it cannot be converted
+    try {
+      JSONObject jsonContainer = new JSONObject(json);
+      text = jsonContainer.getString("text");
+    } catch (JSONException exception) {
+      String errorMessage = "Could not convert text sent to server from JSON.";
+
+      handleError(response, HttpServletResponse.SC_BAD_REQUEST, errorMessage);
+      return;
+    }
+
+    // check that text is valid (not null or empty)
+    if (text == null || text.equals("")) {
+      handleError(response, HttpServletResponse.SC_BAD_REQUEST, "Text input was null or empty");
+      return;
+    }
+
+    // as we want to display the values as well as the decision in our demo
+    // we need to access internal code that the MVP would not
+    PerspectiveAPIFactory factory;
+    
+    try {
+      factory = new PerspectiveAPIFactoryImpl();
+    } catch (ClassNotFoundException | NoSuchMethodException | IllegalAccessException | InvocationTargetException exception) {
+      // if any errors were thrown for looking for api key, send that it was not retrieved back to page
+      handleError(response, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Could not retrieve the Perspective API key.");
+      return;
+    }
```

Comment:
I like how you're avoiding nesting try/catch blocks like this.

Reply:
Thank you!